packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -3,18 +3,68 @@
 // found in the LICENSE file.
 
 import 'dart:math' show min, max;
-import 'dart:ui' as ui show Paragraph, ParagraphBuilder, ParagraphConstraints, ParagraphStyle;
+import 'dart:ui' as ui show Paragraph, ParagraphBuilder, ParagraphConstraints, ParagraphStyle, Rect, Size, PlaceholderAlignment;
 
 import 'package:flutter/foundation.dart';
 import 'package:flutter/gestures.dart';
 import 'package:flutter/services.dart';
 
 import 'basic_types.dart';
+import 'inline_span.dart';
 import 'strut_style.dart';
 import 'text_span.dart';
 
 export 'package:flutter/services.dart' show TextRange, TextSelection;
 
+/// Holds the [Size] and baseline required to represent the dimensions of
+/// a placeholder in text.
+///
+/// Placeholders specify an empty space in the text layout, which is used
+/// to later render arbitrary inline widgets into defined by a [WidgetSpan].
+///
+/// See also:
+///
+///  * [WidgetSpan] a subclass of [TextSpan] that represents an inline widget
+///    embedded within text. The space this widget takes is indicated by a
+///    placeholder.
+///  * [TextSpan] specifies the content. [WidgetSpan]s may be inserted as leaf
+///    nodes within a [TextSpan] tree.
+///  * [RichText] text widget that supports text inline widgets.
+class PlaceholderDimensions {
+  PlaceholderDimensions({
+    @required this.size,
+    @required this.alignment,
+    this.baseline,
+    this.baselineOffset
+  });
+
+  /// Width and height dimensions of the placeholder.
+  Size size;
+
+  /// How to align the placeholder with the text.
+  ui.PlaceholderAlignment alignment;
+
+  /// Distance of the alphabetic baseline from the upper edge of the placeholder.
+  ///
+  /// Only used when [alignment] is [ui.PlaceholderAlignment.baseline].
+  double baselineOffset;
+
+  /// The [TextBaseline] to align to. Used with:
+  ///  * [ui.PlaceholderAlignment.baseline]
+  ///  * [ui.PlaceholderAlignment.aboveBaseline]
+  ///  * [ui.PlaceholderAlignment.underBaseline]
+  ///  * [ui.PlaceholderAlignment.middle]
```

Comment:
i believe you need a blank line before the list starts or it'll all become one paragraph